lib/galaxy/webapps/galaxy/services/tools.py

```diff
@@ -0,0 +1,326 @@
+import logging
+import shutil
+import tempfile
+from json import dumps
+from typing import (
+    Any,
+    Dict,
+    List,
+    Optional,
+    Union,
+)
+
+from starlette.datastructures import UploadFile
+
+from galaxy import (
+    exceptions,
+    util,
+)
+from galaxy.config import GalaxyAppConfiguration
+from galaxy.managers.collections_util import dictify_dataset_collection_instance
+from galaxy.managers.context import (
+    ProvidesHistoryContext,
+    ProvidesUserContext,
+)
+from galaxy.managers.histories import HistoryManager
+from galaxy.model import PostJobAction
+from galaxy.schema.fetch_data import (
+    FetchDataFormPayload,
+    FetchDataPayload,
+    FilesPayload,
+)
+from galaxy.security.idencoding import IdEncodingHelper
+from galaxy.tools import Tool
+from galaxy.tools.search import ToolBoxSearch
+from galaxy.webapps.galaxy.services.base import ServiceBase
+from ._fetch_util import validate_and_normalize_targets
+
+log = logging.getLogger(__name__)
+
+
+class ToolsService(ServiceBase):
+    def __init__(
+        self,
+        config: GalaxyAppConfiguration,
+        toolbox_search: ToolBoxSearch,
+        security: IdEncodingHelper,
+        history_manager: HistoryManager,
+    ):
+        super().__init__(security)
+        self.config = config
+        self.toolbox_search = toolbox_search
+        self.history_manager = history_manager
+
+    def _create_fetch(
```

Comment:
The service methods that are going to be used by the API controller (somewhat *public*) should probably be declared without `_`, not a big deal anyway.